src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/AnalyzePackageStatus.cs

```diff
@@ -95,13 +95,20 @@ private static IReadOnlyCollection<string> ExtractAnalysisResult(IDependencyAnal
                 GetResults("Framework Reference to ", " needs to be deleted ", analysisState.FrameworkReferences.Deletions);
                 GetResults("Framework Reference to ", " needs to be added ", analysisState.FrameworkReferences.Additions);
 
-                void GetResults<T>(string name, string action, IReadOnlyCollection<T> collection)
+                void GetResults<T>(string name, string action, IReadOnlyCollection<Operation<T>> collection)
                 {
                     if (collection.Any())
                     {
                         foreach (var s in collection)
                         {
-                            results.Add(string.Concat(name, s, action));
+                            if (s.Action != null && s.Action.Details != null && s.Action.Details.Any() && !results.Intersect(s.Action.Details).Any())
```

Comment:
Why so many null checks? Are these non-nullable?

Looks like you want to merge these, right? why not make the collection a hashset that will handle duplicates?

Reply:
yes they are all non-nullable.